### PR TITLE
Looking to remove some redundant and unused functions from the ember attribute storage/table

### PR DESCRIPTION
**Pull request by andy31415**
Changes:
  - replace emberAfReadAttribute with emAfReadAttribute since they are the same. removed dataType output argument as it was never used
  - removed use of `emberAfAllowNetworkWriteAttributeCallback` 
      - :warning:  **NOTE: This may be an ABI difference**. Asked in slack if this seems ok - I believe it is as comments talk about ZCL rather than matter
  - removed the justTest argument from emAfAttributeWrite as it was never used

### CODE REVIEW

File: src/app/util/attribute-table.cpp [hunk 1]
```diff
@@ -46,33 +46,13 @@ using namespace chip;
 EmberAfStatus emberAfWriteAttributeExternal(EndpointId endpoint, ClusterId cluster, AttributeId attributeID, uint8_t * dataPtr,
                                             EmberAfAttributeType dataType)
 {
-    EmberAfAttributeWritePermission extWritePermission =
-        emberAfAllowNetworkWriteAttributeCallback(endpoint, cluster, attributeID, dataPtr, dataType);
```
Comment (bzbarsky-apple) [inline src/app/util/attribute-table.cpp:50]: This was supposed to be a place where applications can hook over-the-network writes and allow/deny them, no?  Why are we removing that?  Are we sure applications do not use it?

File: src/app/util/attribute-table.cpp [hunk 2]
```diff
@@ -46,33 +46,13 @@ using namespace chip;
...
 {
-    return emAfWriteAttribute(endpoint, cluster, attributeID, dataPtr, dataType,
-                              true,   // override read-only?
-                              false); // just test?
-}
-
-EmberAfStatus emberAfReadAttribute(EndpointId endpoint, ClusterId cluster, AttributeId attributeID, uint8_t * dataPtr,
-                                   uint16_t readLength)
-{
-    return emAfReadAttribute(endpoint, cluster, attributeID, dataPtr, readLength, nullptr);
```
Comment (bzbarsky-apple) [inline src/app/util/attribute-table.cpp:75]: The "emAf" bits are basically not public API.  The "emberAf" bits are.  That's why there was the forwarding going on there....

File: src/app/util/attribute-table.cpp (hunk 1 above)
Comment (andy31415) [reply src/app/util/attribute-table.cpp:50]: This is the bit where I was unsure ... however I have seen no use of it and it uses an ugly hook of using weak-ptr. I am looking to have some path to uncouple things. If we do have users of this I will see how I can make it pluggable instead (and keep it). 

Absolutely trying to figure out if adding this was a violation of `you ain't gonna need it` or not.

File: src/app/util/attribute-table.cpp (hunk 2 above)
Comment (andy31415) [reply src/app/util/attribute-table.cpp:75]: I was unclear why because one was just calling the other with identical parameters.

If this is a naming convention, I can rename this to emberAfReadAttribute instead of `emAf...`. Are you thinking that is sufficient or do you imagine the 'public version' will eventually do more?

For background of where I am going with these changes: looking to make the API as small as we use and then I would like to place it into something replaceable via an interface, to get testability for clusters without requiring custom zap-regen for every test app.  This is how I got to touching read/write of attributes overall.

File: src/app/util/attribute-table.cpp (hunk 1 above)
Comment (tcarmelveilleux) [reply src/app/util/attribute-table.cpp:50]: Well, it's the ACL that is supposed to be the source of truth for access. Having a random block from a callback, is not something that is expected. We don't have extensive tests to guard against arbitrary behaviors there in cert!

Comment (bzbarsky-apple) [reply src/app/util/attribute-table.cpp:50]: ACL cannot implement checks based on the actual value being written, which this can.

Now in theory all such checks should be defined by the spec and hence should be implemented in our cluster code and hence applications should not need to do anything here.  The question is whether this theory is correct.

I personally don't mind removing this hook; the question is whether it breaks anyone.

File: src/app/util/attribute-table.cpp (hunk 2 above)
Comment (bzbarsky-apple) [reply src/app/util/attribute-table.cpp:75]: > If this is a naming convention, I can rename this to emberAfReadAttribute instead of emAf...

And move the declaration into the "more public" af.h header?  That would be fine from my point of view.

Comment (andy31415) [reply src/app/util/attribute-table.cpp:75]: For what its worth, I hate it when the header name and impl name differ :/ Will not fix here since I plan to add a replaceable interface/shim and  anyway.

Comment (andy31415) [reply src/app/util/attribute-table.cpp:75]: I will move emberAfWriteAttributeExternal too then.

File: src/app/util/attribute-table.cpp (hunk 1 above)
Comment (andy31415) [reply src/app/util/attribute-table.cpp:39]: I was basing my removal on what seemed zigbee legacy, specifically the comment bits of:

```
...
 Note:   The ZCL specification does not (as of this
 * writing) specify any permission-level security for writing writeable
 * attributes. As far as the ZCL specification is concerned, if an attribute is
 * writeable, any device that has a link key for the device should be able to
 * write that attribute. Furthermore if an attribute is read only, it should not
 * be written over the air. Thus, if you implement permissions for writing
 * attributes as a feature, you MAY be operating outside the specification. This
 * is unlikely to be a problem for writing read-only attributes, but it may be a
 * problem for attributes that are writeable according to the specification but
 * restricted by the application implementing this callback.
 ```
 
 Which does not seem to apply to matter. However it is also true that someone may have decided to implement this for some reason (not sure why they would though...). I am trying to see if sufficient slack communication will find someone that used it or take silence as an indicator of "not used". Unfortunately I can only prove usage if someone answers, however I cannot prove "never used" ...

File: src/app/util/af.h [hunk 3]
```diff
@@ -101,10 +101,23 @@ bool emberAfContainsClient(chip::EndpointId endpoint, chip::ClusterId clusterId)
...
+ * TODO: this not checking seems off - what if this is run without Accessors.h ?
  */
 EmberAfStatus emberAfWriteAttribute(chip::EndpointId endpoint, chip::ClusterId cluster, chip::AttributeId attributeID,
                                     uint8_t * dataPtr, EmberAfAttributeType dataType);
 
+/**
+ * Write an attribute for a request arriving from external sources.
+ *
+ * This will check attribute attribute writeability and that
+ * the privided data type matches the expected data type.
```
Comment (bzbarsky-apple) [inline src/app/util/af.h:116]:
```suggestion
 * This will check attribute writeability and that
 * the provided data type matches the expected data type.
```

File: src/app/zap-templates/templates/app/attributes/Accessors-src.zapt [hunk 4]
```diff
@@ -10,8 +10,8 @@
 #include <app-common/zap-generated/attribute-type.h>
 #include <app-common/zap-generated/ids/Attributes.h>
 #include <app-common/zap-generated/ids/Clusters.h>
-#include <app/util/af.h>
 #include <app/util/attribute-storage-null-handling.h>
+#include <app/util/af.h>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/attributes/Accessors-src.zapt:14]: Why this change?

File: src/app/util/af.h [hunk 5]
```diff
@@ -101,10 +101,23 @@ bool emberAfContainsClient(chip::EndpointId endpoint, chip::ClusterId clusterId)
...
+ */
+EmberAfStatus emberAfWriteAttributeExternal(chip::EndpointId endpoint, chip::ClusterId cluster, chip::AttributeId attributeID,
```
Comment (bzbarsky-apple) [inline src/app/util/af.h:118]: This isn't really public API, right?  As in, applications should not be calling this, ever.  Why did this declaration get moved here?

Comment (andy31415) [reply src/app/util/af.h:118]: because it starts with `ember` in the previous version, so I moved it here. I am not familiar enough with ember right now, so my instinct was: `emAf` are internal, `emberAf` are external. Are there any conventions in this codebase that I can follow or at least start enforcing?

To me the layout is generally not ideal: having a separate header that exposes things define in different cpp files is not ideal. We should have had `af.cpp`.

File: src/app/zap-templates/templates/app/attributes/Accessors-src.zapt (hunk 4 above)
Comment (andy31415) [reply src/app/zap-templates/templates/app/attributes/Accessors-src.zapt:14]: I had removed it and addeed back, this re-ordered. Will move it back.

File: src/app/util/af.h (hunk 5 above)
Comment (andy31415) [reply src/app/util/af.h]: Moved back to old location, but renamed as `emAfWriteAttributeExternal` to try to have more of `emberAf` is public and `emAf` is internal.... 

Overall I believe apps should generally never call `em*` at all and should instead use Accessors.h and similar - that is the route that I am trying to move towards.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by tcarmelveilleux — APPROVED** [review]
